### PR TITLE
feat(NODE-5188): add alternative runtime detection to client metadata

### PR DESCRIPTION
**Pull request by nbbeeken**
#### What is changing?

Changes platform metadata based on globals in the driver's runtime environement.

#### What is the motivation for this change?

Node.js compatibility in Deno and Bun have made it possible to run the driver, it is useful to denote the difference between the runtimes in the platform information.

### Double check the following

- [x] Ran `npm run check:lint` script
- [x] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [x] PR title follows the [correct format](https://www.conventionalcommits.org/en/v1.0.0/): `type(NODE-xxxx)[!]: description`
  - Example: `feat(NODE-1234)!: rewriting everything in coffeescript`
- [x] Changes are covered by tests
- [x] New TODOs have a related JIRA ticket

### CODE REVIEW

**Comment by nbbeeken** [comment]
Server log lines:
```json
{"t":{"$date":"2023-04-12T15:53:58.738-04:00"},"s":"I",  "c":"NETWORK",  "id":51800,   "ctx":"conn6","msg":"client metadata","attr":{"remote":"127.0.0.1:64794","client":"conn6","doc":{"driver":{"name":"nodejs","version":"5.2.0"},
"platform":"Deno v1.32.3, LE",
"os":{"name":"darwin","architecture":"arm64","version":"22.4.0","type":"Darwin"}}}}

{"t":{"$date":"2023-04-12T16:03:31.216-04:00"},"s":"I",  "c":"NETWORK",  "id":51800,   "ctx":"conn18","msg":"client metadata","attr":{"remote":"127.0.0.1:59421","client":"conn18","doc":{"driver":{"name":"nodejs","version":"5.2.0"},
"platform":"Bun v0.5.4, LE",
"os":{"name":"darwin","architecture":"arm64","version":"22.4.0","type":"Darwin"}}}}
```

**Review by alexbevi — APPROVED** [review]
LGTM

File: src/cmap/handshake/client_metadata.ts [hunk 1]
```diff
@@ -234,3 +234,58 @@ export function getFAASEnv(): Map<string, string | Int32> | null {
...
+/**
+ * @internal
+ * Get current JavaScript runtime platform
+ *
+ * NOTE: The version information fetching is intentionally written aggressively defensive
+ * to avoid having a released driver version that becomes incompatible
+ * with a future change to these global objects
+ */
+function getRuntimeInfo(): string {
+  if ('Deno' in globalThis) {
```
Comment (durran) [inline src/cmap/handshake/client_metadata.ts:263]: Why are all the checks necessary? Can't we assume if `Deno` or `Bun` are present in the global namespace that they conform to the type we expect? And that if a user was to randomly put them there all bets are off?

**Review by durran — CHANGES_REQUESTED** [review]

File: src/cmap/handshake/client_metadata.ts [hunk 2]
```diff
@@ -234,3 +234,58 @@ export function getFAASEnv(): Map<string, string | Int32> | null {
...
+    const version =
+      Deno != null &&
+      typeof Deno === 'object' &&
+      'version' in Deno &&
+      Deno.version != null &&
+      typeof Deno.version === 'object' &&
+      'deno' in Deno.version &&
+      typeof Deno.version.deno === 'string'
+        ? Deno.version.deno
+        : '0.0.0';
```
Comment (nbbeeken) [inline src/cmap/handshake/client_metadata.ts:273]: Add `-unknown`

File: src/cmap/handshake/client_metadata.ts [hunk 3]
```diff
@@ -234,3 +234,58 @@ export function getFAASEnv(): Map<string, string | Int32> | null {
...
+ * with a future change to these global objects
+ */
+function getRuntimeInfo(): string {
+  if ('Deno' in globalThis) {
+    const version =
+      Deno != null &&
+      typeof Deno === 'object' &&
+      'version' in Deno &&
+      Deno.version != null &&
+      typeof Deno.version === 'object' &&
```
Comment (nbbeeken) [inline src/cmap/handshake/client_metadata.ts:269]: use optional chaining

File: src/cmap/handshake/client_metadata.ts (hunk 1 above)
Comment (nbbeeken) [reply src/cmap/handshake/client_metadata.ts:263]: We just want to write code that won't end up throwing if there's a potential change. We can reduce the checks here to optional chains though

**Review by baileympearson — APPROVED** [review]

File: src/cmap/handshake/client_metadata.ts [hunk 4]
```diff
@@ -234,3 +234,39 @@ export function getFAASEnv(): Map<string, string | Int32> | null {
...
+ * @internal
+ * Get current JavaScript runtime platform
+ *
+ * NOTE: The version information fetching is intentionally written defensively
+ * to avoid having a released driver version that becomes incompatible
+ * with a future change to these global objects.
+ */
+function getRuntimeInfo(): string {
+  if ('Deno' in globalThis) {
+    const version = typeof Deno?.version?.deno === 'string' ? Deno?.version?.deno : '0.0.0-unknown';
```
Comment (durran) [inline src/cmap/handshake/client_metadata.ts:260]: `typeof new String(‘’)` returns “object” so a safer check: https://github.com/mongodb/node-mongodb-native/pull/3637/files#diff-39b2554fd18da165b59a6351b1aafff3714e2a80c1435f2de9706355b4d32351R69

**Review by durran — CHANGES_REQUESTED** [review]

File: src/cmap/handshake/client_metadata.ts (hunk 4 above)
Comment (nbbeeken) [reply src/cmap/handshake/client_metadata.ts:260]: Is this something we want to support? Generally, none of our code supports string objects / number objects / etc. I would be suspicious of code that uses String instances.

Unrelated but here is how String objects serialize in BSON, just to exemplify the degree to which we assume these constructors are assumed to not be used:
```js
> BSON.deserialize(BSON.serialize({ a: new String('abc') }))
{ a: { '0': 'a', '1': 'b', '2': 'c' } }
```

Comment (durran) [reply src/cmap/handshake/client_metadata.ts:260]: Didn't know if we wanted to or not, given the serialisation example it seems not then, although unexpected. But I guess if no one has ever reported that no one really uses them.

**Review by durran — APPROVED** [review]

**Pull request merged**